packages/adblocker-content/adblocker.ts

```diff
@@ -73,6 +85,8 @@ export function extractFeaturesFromDOM(roots: Element[]): {
   const classes: Set<string> = new Set();
   const hrefs: Set<string> = new Set();
   const ids: Set<string> = new Set();
+  // Try to reduce duplicate entries by using Set instead of an array.
+  const processedElements: Set<Element> = new Set();
```

Comment:
Is this still required if we deduplicate elements from `root`? Did you check that it's possible to have duplicated elements at this point?

Reply:
@remusao I'm checking in depth but lagging always happens with Array. Checked with `Array.from` and spreading operator. I think I'm doing something wrong but cannot find any resources on this. The sample implementation looks like the following:

```ts
      // An array of updated nodes in this timespan (or debounce session).
      // This variable needs to be kept outside of mutation observer callback
      // because the role of this array is to keep and accumulate the elements
      // to be passed as an argument before debouncing completes its lifecycle.
      let nodes: Element[] = [];

      const debouncedHandleUpdatedNodes = debounce(() => {
        const nodesRef = Array.from(nodes);

        // Initialise and assign new object instead of modifying the existing one.
        // Modifying the existing one will impact to the function running after this onComplete callback.
        nodes = [];

        this.handleUpdatedNodes(nodesRef);
      }, 25);
```

Reply:
I'll propose additional changes if I can find some improvements on this, but I'm sorry to say like this at the moment.

Reply:
I've tested it without this Set and there are no performance issues. Would rather see it removed as it must have some memory impact.

Reply:
When I try, the "root" elements are deduplicated (e.g. `for (const root of [...new Set(roots)])` iterates over the same entries). However, the querySelectorAll result sets do overlap. IMHO, the current solution using a Set to mark the visited ones, makes sense.

Reply:
I did a simple testing and got the following numbers regarding this. However, it still doesn't describe the flame graph difference, but now I can get why the performance differs.

<details>
<summary>Details</summary>

![image](https://github.com/ghostery/adblocker/assets/30369714/ce28d43c-f441-4278-b35e-2a3b33700476)

</details>